### DISCUSSION TITLE
AWS deployment using Docker?

### DISCUSSION THREAD

**Discussion by guidocioni** [zauberzeug/nicegui#533] [Q&A]
As a long-term user of Dash, Flask and Bootstrap I don't know why it took me so long to find out about this amazing package.
I loved Dash for prototyping and created so many applications with it but the problem is that it took me always too long to deploy it on cloud architectures like AWS without using a dedicated virtual machine instance.

As `nicegui` has a dedicated Docker image I'd suppose this is far easier to deploy on AWS as a dockerized Lambda. 
Unfortunately the section on the documentation (https://nicegui.io/reference#server_hosting) doesn't really expand on this possibility. 

Did anyone try to do that? 

I think it would also be helpful to have a full repo example with Docker (right now there is only a link to the official image and a docker compose yaml).

**Comment by rodja** [comment]
We have not tried Lambda yet and it may not be the best fit from the AWS product line. Lambda is designed for running short-lived, stateless functions in response to events. NiceGUI on the other hand gains its power from a long living backend service which serves dynamic html/js which is manipulated via websocket.

Here are some suggestions from ChatGPT 4 (not asking about brand new NiceGUI which is unfortunately unknown yet, but rather the underlying FastAPI framework):

1. AWS Elastic Beanstalk: A Platform-as-a-Service (PaaS) that can automatically manage and scale your FastAPI application. You can use a pre-configured Python environment and deploy your application using the Elastic Beanstalk console or CLI.
2. AWS EC2: You can deploy your FastAPI application on an EC2 instance and manually configure the server, networking, and other related components. This option provides more control over the environment but requires more management.
3. AWS Lightsail: A simplified version of EC2 that is designed for smaller projects or developers who want a more straightforward deployment process.
4. Containerized solutions: Package your FastAPI application as a Docker container and deploy it on services like AWS ECS (Elastic Container Service), AWS ECR (Elastic Container Registry), or AWS EKS (Elastic Kubernetes Service).

**Reply by guidocioni** [reply]
Well, although lamdba is not designed to run full fledged web applications, it can indeed run flask applications without paying for ec2 or beanstalk (which does nothing but using a ec2 machine). Have a look here 
https://github.com/zappa/Zappa

The idea is that you only pay for the time when your web application is actually used instead than paying for a machine turned on 24/7. This is actually really good value for simple apps. For production I agree it wouldn't be the ideal solution.

In the past I managed to deploy applications with Zappa and with a docker container through lambda registry, but it was a pain in the ... to maintain it.

I think it should be possible to do the same with nicegui after exposing the right ports and having the lambda execution runtime installed in the container. Anyway it was just a curiosity in case someone already did it 😄

**Reply by Ha44a** [reply]
@guidocioni 

Did you ever get this to work? I am currently trying to implement a similar deployment

**Reply by guidocioni** [reply]
> @guidocioni
> 
> Did you ever get this to work? I am currently trying to implement a similar deployment

No, in the end I continued using Dash/Flask for my apps because I'm too lazy (and old) to learn yet another framework :D

**Reply by Ha44a** [reply]
Yep, fair enough! Cheers.

**Comment by mpechter** [comment]
If anyone is looking for a way to deploy NiceGUI on AWS, a great option is App Runner. However, App Runner doesn't support WebSockets. By forcing Socket.IO to use long polling instead, you can deploy there without issue:

```py
from nicegui import ui, core

core.app.config.socket_io_js_transports = ["polling"]
ui.run(proxy_headers=True, forwarded_allow_ips="*")
```

The proxy_headers and forwarded_allow_ips settings are needed because App Runner terminates TLS and forwards traffic as HTTP — without them, uvicorn doesn't know the client connected over HTTPS.
